### ISSUE TITLE
Debugging

### ISSUE THREAD

**Issue by dandv** [bitinn/node-fetch#284]
Would be great to have a debugging feature to examine requests before they go down the wire, [a la `request`](https://github.com/request/request#debugging).

**Comment by jimmywarting** [comment]
Think this has been brought up before...

**Comment by lostpebble** [comment]
Yep, debugging requests on Node is tough with this module. Would really like a way to inspect / debug things.

**Comment by sabrehagen** [comment]
Yes, this would be a major help and has been requested before #150.

**Comment by bitinn** [comment]
A few suggestions:

1. Be more specific about what you want from *debugging*; we are pretty transparent about exposing errors (unlike Fetch in browser, which doesn't give you much to investigate), but maybe you want more.

2. We are careful not to expose too much control related to native HTTP(S) module, because at that point you might as well be working with native modules. Nevertheless, we like to hear your pain point.

3. We are very understaffed (see #252 and [our contributors graph](https://github.com/bitinn/node-fetch/graphs/contributors)), if you got a great idea, open an PR or issue, it starts the conversation.

4. There are literally PR opened related to debugging (see #363 #364), but we hear nothing from people about if they want it or not.

**Comment by lostpebble** [comment]
Sorry for not being specific enough.

Basically I came here looking for debugging options because my Authentication headers were not working. Turns out I was using the wrong syntax (thanks to me trying to work around incorrect TypeScript definitions https://github.com/Microsoft/TypeScript/issues/19656 - which had appeared to be working at first):

```
headers: [["Content-Type", "application/json"], ["Authorization", `Basic ${Buffer.from(`vibescout:${ServiceConfig.MailChimp.apiToken}`).toString("base64")}`]]
```
Instead of:
```
headers: {
    "Content-Type": "application/json",
    Authorization: `Basic ${Buffer.from(`vibescout:${ServiceConfig.MailChimp.apiToken}`).toString("base64")}`,
}
```

But what would have been helpful to debug it was if the final / created request being sent out, including headers etc. was able to be printed to console - then I could explore it with Google Chrome devtools and see that my headers were not being created properly somehow. Instead I was left wondering if my API token had expired.

So my specification for basic debugging would be a switch to say: print the request object and the response object to the debug console for further exploration.

Understandably you don't seem to have enough time to implement every little thing now. I'll see if I can dig in and help out at some point.

**Comment by jimmywarting** [comment]
Just something that pop in to my mind is: wouldn't it be cool if you could see network request in chrome's devtool?

you can run node with --inspect flag and debug your application but i don't know if you can see network request or not...

**Comment by lostpebble** [comment]
@jimmywarting exactly. I wish that was the case - the same as we have for frontend JavaScript code, you can click the networks tab and view all the exact details about all the requests. The second best solution would be a debugging log to the console.

**Comment by TimothyGu** [comment]
> But what would have been helpful to debug it was if the final / created request being sent out, including headers etc. was able to be printed to console - then I could explore it with Google Chrome devtools and see that my headers were not being created properly somehow. 

It has been discussed in Node.js to make the builtin http library hook up to DevTools, but the groundwork is still being laid for non-V8 inspector domains (see e.g. https://github.com/nodejs/node/issues/16627), so it's going to take a while.

I'd be open to an option in node-fetch to save all traffic into a HAR, just like [request-capture-har](https://github.com/paulirish/request-capture-har).

**Comment by jimmywarting** [comment]
Oh, interesting!
...about built in http to devtool, can't wait

**Comment by mr-beerkiss** [comment]
Chrome dev tools are not really useful if you are running Node.js apps in production that use fetch as their HTTP client.  Ideally you want to be able to retrieve metrics from each request you make so that you can log them and apply your own monitoring on top.  #364 attempts to provide some of the common timings your might see you Chrome dev tools.  There metrics are particularly useful if you are debugging network issues inside your production environment, for example slow DNS.

I can see the argument of wanting to align closely to the spec.  Perhaps if we emit events for apps to hook into we could accomplish a similar thing?

**Comment by bitinn** [comment]
@mr-beerkiss Yeah I can see why this is useful in production. And really this is one of the features people have requested before.

It's tempting to simply merge such PR but it also means in future people would ask us to insert more features that native HTTP module provides.

The alternative: exposing socket events or emit custom events are also something people have talked about...

Again, I am not against giving people useful information to log and debug, but it seems like **the ultimate resolution has always been to expose the underlying HTTP request/response object**. That's not the philosophy of Fetch, but if enough people want it so badly, we might also give people the power (use at your own risk).

PS: @TimothyGu is the owner for v2 branch, so he has the final say on this. **To me, exposing req/res objects are not great, but it's better than any other elegant solution we can come up with yet have no time to implement...**

**Comment by bitinn** [comment]
And for @lostpebble, if we can somehow expose underlying HTTP req/res object, does it give you a path to do what you want?

**Comment by jimmywarting** [comment]
> insert more features that native HTTP module provides.

Now we don't want that... Want it to stay as close to spec as possible and be the bear minimum.
But I do however wish for more feature to be available like cache, cookie jar, debugging, proxies and other useful features that isn't part of the spec but are still useful

So here is my suggestion: make it so that it's possible to extend the functionality of node-fetch in some modular way with 3th party modules

**Comment by bitinn** [comment]
> make it so that it's possible to extend the functionality of node-fetch in some modular way with 3th party modules

In the case above, timing info is dependent on hooking into the socket event. We have to expose it somehow or collect these info on our own. Do you see a better approach?

**Comment by jimmywarting** [comment]
Don't know how think we need to make a new issues that are more focused on how to make this modular in a way that it's possible to to all things necessary

Whether it will be by 
- exposing underlying http req/res in the Request/Response class
- or make it event driven
- #229 had some idea about building wrappers

**Comment by lostpebble** [comment]
@bitinn Yes, if I can somehow explore the underlying request object just before it gets sent out, headers and all that stuff - that would be what I was looking for coming into this issue. The response object as well, although it was less of a priority - as I was more looking to debug what I was doing wrong (sending out incorrect requests).

**Comment by mr-beerkiss** [comment]
Having underlying request/response would mean we can solve our monitoring problems in our local domain so that would work for us.  

How would this api look though?

**Comment by jimmywarting** [comment]
Maybe this is becoming more of a interest: https://www.npmjs.com/package/fetch-event?

**Comment by bitinn** [comment]
@lostpebble @mr-beerkiss 

One possibility (if all you want is to expose req information and not a plugin system):

- For request debug: a custom event that emit the request options just before we send it out.
- For request logging: a custom event that emit the actual request options as we send it out.

But it should be read-only, I am not going to help people troubleshoot HTTP request problems due to messing with request object. (Fetch is really designed to keep you from doing that).

https://github.com/bitinn/node-fetch/blob/master/src/index.js#L50

**Comment by mr-beerkiss** [comment]
@bitinn Agree req should be read only.  Not sure if that is enforceable though? 

@jimmywarting that seems like a good approach of emit events from the fetch function, although I don't think you should be able to interfere with the underlying mechanics of fetch (eg. in the code example you are sending a new response via the event)

**Comment by jimmywarting** [comment]
I tried to mimic the service workers event mechanism where you actually are able to respond with a new response using [respondWith](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/respondWith)

**Comment by mr-beerkiss** [comment]
Ah fair enough, I haven't played around with Service workers and clearly I should!

**Comment by pimterry** [comment]
I hit this problem, and I've built a standalone open-source tool to fix it. It's called HTTP Toolkit - it automatically intercepts all HTTP(S) from node-fetch, the built-in `http` module, and browser JS too, so you can inspect, debug & rewrite all of it in one place.

Built-in support would be nice of course, but this might help some of you in the meantime! Details are over here: https://httptoolkit.tech/mock/javascript/.

**Comment by NotMoni** [comment]
I think that solves the problem for now.
